src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -385,10 +383,299 @@ public static void EnsureSocketsAreInitialized()
+
+                if (errorCode != SocketError.IOPending)
+                    ProcessResult(errorCode, context);
+            }
+        }
+
+        private static unsafe void GetAddressInfoExCallback([In] int error, [In] int bytes, [In] NativeOverlapped* overlapped)
+        {
+            try
+            { }
+            finally
+            {
+                // Can be casted directly to QueryContext* because the overlapped is its first field
+                GetAddrInfoExContext* context = (GetAddrInfoExContext*)overlapped;
+
+                ProcessResult((SocketError)error, context);
+            }
+        }
+
+        private static unsafe void ProcessResult(SocketError errorCode, GetAddrInfoExContext* context)
+        {
+            GetAddrInfoExState state = context->GetQueryState();
+
+            if (state == null || !state.SetCallbackStartedOrCanceled())
+                return;
+
+            try
+            {
+                if (errorCode != SocketError.Success)
+                {
+                    state.CompleteAsyncResult(new SocketException((int)errorCode));
+                    return;
+                }
+
+                AddressInfoEx* result = context->Result;
+                string canonicalName = null;
+
+                List<IPAddress> addresses = new List<IPAddress>();
+
+                while (result != null)
+                {
+                    if (canonicalName == null && result->ai_canonname != IntPtr.Zero)
+                        canonicalName = Marshal.PtrToStringUni(result->ai_canonname);
+
+                    IPAddress ipAddress = null;
+
+                    if (result->ai_family == AddressFamily.InterNetwork)
+                    {
+                        ipAddress = CreateIPv4Address(result->ai_addr, result->ai_addrlen);
+                    }
+                    else if (SocketProtocolSupportPal.OSSupportsIPv6 && result->ai_family == AddressFamily.InterNetworkV6)
+                    {
+                        ipAddress = CreateIPv6Address(result->ai_addr, result->ai_addrlen);
+                    }
+
+                    if (ipAddress != null)
+                        addresses.Add(ipAddress);
+
+                    result = result->ai_next;
+                }
+
+                if (canonicalName == null)
+                    canonicalName = state.HostAddress;
+
+                state.CompleteAsyncResult(new IPHostEntry
+                {
+                    HostName = canonicalName,
+                    Aliases = Array.Empty<string>(),
+                    AddressList = addresses.ToArray()
+                });
+            }
+            finally
+            {
+                state.Dispose();
+            }
+        }
+
+        private static unsafe IPAddress CreateIPv4Address(byte* socketAddress, int addressLength)
```

Comment:
Could we write these Create methods in terms of spans rather than pointers and have the call site create the span? e.g.
```C#
private static IPAddress CreateIPv4Address(Span<byte> socketAddress) =>
    new IPAddress(socketAddress.Slice(4, IPAddressParserStatics.IPv4AddressBytes));
```
and then at the call site like:
```C#
Span<byte> addressSpan = new Span<byte>(result->ai_addr, result->ai_addrlen);
if (pAddressInfo->ai_family == AddressFamily.InterNetwork)
{
    if (addressSpan.Length == IPv4SocketAddressSize) addresses.Add(CreateIPv4Address(addressSpan));
}
else if (pAddressInfo->ai_family == AddressFamily.InterNetworkV6)
{
    if (addressSpan.Length == IPv6SocketAddressSize) addresses.Add(CreateIPv6Address(addressSpan));
}
```
Etc.